FuelLabs/fuel-indexer
# enhancement: WASM error codes, early exit on error or kill switch

### Description

UPDATE: in addition to adding error codes to host functions as described below, this PR adds another host function:

```rust
fn early_exit(err_code: WasmIndexerError) -> !
```

Using this function we can remove more `unwrap` or `expect` calls.

Building on https://github.com/FuelLabs/fuel-indexer/pull/1293.

This PR follows the example in `wasmer` repository:

https://github.com/wasmerio/wasmer/blob/master/examples/early_exit.rs#L60

```
    fn early_exit() -> Result<(), ExitCode> {
        // This is where it happens.
        Err(ExitCode(1))
    }
```

Note that the return value is `Result<(), ExitCode>` while the function is imported in WASM as

https://github.com/wasmerio/wasmer/blob/master/examples/early_exit.rs#L41

```
  (type $early_exit_t (func (param) (result)))
```

That is, as if it were `fn early_exit();`.

This PR does the same for our FFI functions. For instance:

```rust
fn put_object(
    mut env: FunctionEnvMut<IndexEnv>,
    type_id: i64,
    ptr: u32,
    len: u32,
) -> Result<(), WasmIndexerError> { }
```

While their types—as declared to the WASM module which will use them—doesn't mention the `Result` value:

https://github.com/FuelLabs/fuel-indexer/pull/1337/files#diff-447e55cb0fae7d02ac8fd8f7de5a521cbf60d3f2422b5166c39faeb9f995bea6R21-R29

The one difference between how things are handled in `early_exit` example versus this PR is getting the error code out:

https://github.com/wasmerio/wasmer/blob/master/examples/early_exit.rs#L93

When I tried the same method:
```rust
Err(e) => match e.downcast::<ExitCode>()
```

I got no error out of it. However, the error was printed as `RuntimeError(User(EarlyExit))`, so the error was handled correctly. (`User` indicates that a user-defined error triggered a WASM trap).

So, I ended up handling the error this way:

https://github.com/FuelLabs/fuel-indexer/pull/1337/files#diff-fa722c66d2896c684e5a26aa2c56b7ec1734888609fbb64ecbc7dfd6a420c4ffR933

```rust
            } else {
                if let Some(e) = e
                    .source()
                    .and_then(|e| e.downcast_ref::<WasmIndexerError>())
                {
                    error!("Indexer({uid}) WASM execution failed: {e}.");
                } else {
                    error!("Indexer({uid}) WASM execution failed: {e:?}.");
                };
                self.db.lock().await.revert_transaction().await?;
                return Err(IndexerError::from(e));
            }
```

I am unsure why I couldn't `downcast` the error the same way as was shown in the example.

### Testing steps

Manual testing:

1. Start an indexer.

```
cargo run -p fuel-indexer -- run --fuel-node-host beta-4.fuel.network --fuel-node-port 80 --replace-indexer --manifest examples/fuel-explorer/fuel-explorer/fuel_explorer.manifest.yaml
```

2. Redeploy:

```
cargo run -p forc-index -- deploy --path examples/fuel-explorer/fuel-explorer --replace-indexer
```

Expected output:

```
2023-09-07T09:12:49.663539Z  INFO fuel_indexer::service: 399: Resuming Indexer(fuellabs.explorer) from block 1240
2023-09-07T09:12:54.321165Z  INFO fuel_indexer::database: 206: Database loading schema for Indexer(fuellabs.explorer) with Version(143093c6bbe3a8937f4fb71514cbe6799266c44398866dbbd23e12b977bc1641).
2023-09-07T09:12:54.346555Z  INFO fuel_indexer::executor: 110: Indexer(fuellabs.explorer) subscribing to Fuel node at beta-4.fuel.network:80
2023-09-07T09:12:54.346681Z  WARN fuel_indexer::executor: 117: No end_block specified in manifest. Indexer will run forever.
2023-09-07T09:12:54.346717Z  INFO fuel_indexer::service: 335: Indexer(fuellabs.explorer) was replaced. Stopping previous version of Indexer(fuellabs.explorer).
2023-09-07T09:12:54.347376Z ERROR fuel_indexer::executor: 941: Indexer(fuellabs.explorer) WASM execution failed: Kill switch has been triggered.
2023-09-07T09:12:54.347819Z  INFO fuel_indexer::executor: 199: Kill switch flipped, stopping Indexer(fuellabs.explorer). <('.')>
```

In particular, this line indicates that an early termination happened due to the kill switch:

```
2023-09-07T09:12:54.347376Z ERROR fuel_indexer::executor: 941: Indexer(fuellabs.explorer) WASM execution failed: Kill switch has been triggered.
```

### Changelog

* FFI functions can access the kill switch indicator and exit early when the kill switch has been triggered.
* FFI functions can exit early on error, including `sqlx` database operation errors.
* Add WASM error codes (from @deekerno's PR)
* Get an error code from the call to the indexer's WASM module.
